cmd/kubelet/app/server.go

```diff
@@ -224,7 +224,7 @@ HTTP server: The kubelet can also listen for HTTP and respond to a simple API
 			}
 
 			// Config and flags parsed, now we can initialize logging.
-			logs.InitLogs()
+			logs.InitLogs(logs.FeatureGate(utilfeature.DefaultFeatureGate))
 			logOption := &logs.Options{Config: kubeletConfig.Logging}
 			if err := logOption.ValidateAndApply(); err != nil {
```

Comment:
`logOption.ValidateAndApply()` seems like the right place to pass a feature gate registry that will be used to make sure any non-GA fields in the config have the corresponding feature enabled.

The flow I was envisioning was:

```
// caller registers feature gates for components it is using
logs.AddFeatureGates(utilfeature.DefaultFeatureGate)
leaderelection.AddFeatureGates(utilfeature.DefaultFeatureGate)

// caller sets up component options, either by embedding component configs in their own config file, or by binding CLI flags
logOptions := logs.Options{Config: myConfig.Logging}
leaderElectionOptions := leaderelection.Options{Config: myConfig.LeaderElection}
- or -
logOptions := logs.NewOptions()
logOptions.AddFlags(...)
leaderElectionOptions := leaderelection.NewOptions()
leaderElectionOptions.AddFlags(...)

... caller loads their config file / parses flags / initializes feature gates, etc ...

// caller validates/applies component config, informed by enabled features
logOptions.ValidateAndApply(utilfeature.DefaultFeatureGate)
leaderElectionOptions.ValidateAndApply(utilfeature.DefaultFeatureGate)
```

That limits all feature gate use to setup time.

Reply:
Not all binaries use `ValidateAndApply`. Some only use klog + `logs.InitLogs`. If we don't change `InitLogs`, then those binaries will use klog's default for contextual logging, which is "enabled" because in klog's API we don't have "alpha quality, off by default" features.

Having said that, I think that's okay. Those binaries which only use klog + `logs.InitLogs` probably also don't have a contextual logger - "probably" because they might set their own one instead of the JSON logger from component-base. There is no known binary which does that.

@liggitt : do you agree that changing `ValidateAndApply` is a reasonable compromise between "complexity of the API" and "completeness"?

Also, I take it that you want the caller to be responsible for registering feature gates, i.e. let component-base/logs use FeatureGate.Enabled and panic when the caller got it wrong? Just double-checking. I'm okay with that, too, because now it affects a smaller number of callers and those can be adapted.

Reply:
> Some only use klog + logs.InitLogs. If we don't change InitLogs, then those binaries will use klog's default for contextual logging, which is "enabled" because in klog's API we don't have "alpha quality, off by default" features.

my primary concern is k8s components. whatever mechanism they use should support opt-in of features we consider alpha


> There is no known binary which does that.

As long as all k8s.io/kubernetes components are consistent in default enablement, that seems reasonable to me

> @liggitt : do you agree that changing ValidateAndApply is a reasonable compromise between "complexity of the API" and "completeness"?

I want all the paths component-base supports to init/configure logging to require opt in to alpha logging features. If the only path it supports is ValidateAndApply, then that's fine. If it also has other paths (CLI flags, etc) that bypass ValidateAndApply then those should check feature opt-in as well. 

> Also, I take it that you want the caller to be responsible for registering feature gates, i.e. let component-base/logs use FeatureGate.Enabled and panic when the caller got it wrong? Just double-checking. I'm okay with that, too, because now it affects a smaller number of callers and those can be adapted.

yeah, and if that check is always done at setup time, we won't hit situations where some later runtime log call triggers a feature gate check beyond startup

Reply:
> be used to make sure any non-GA fields in the config

Hmm, I missed that part. I need to point out that `ContextuaLogging` is not about fields in `LoggingConfiguration` - that is what `LoggingAlphaOptions` and `LoggingBetaOptions` in https://github.com/kubernetes/kubernetes/pull/105797 cover. For those, `ValidateAndApply` is the right function and what that other PR uses.

`ContextuaLogging` controls the behavior of klog's `FromContext`, `LoggerWithValues` and `LoggerWithName` and `NewContext`, that's why it is kind of relevant also without `LoggingConfiguration`. As I said in my previous comment, the question just is whether we really want or need to cover those binaries which don't use `LoggingConfiguration`.

Reply:
Had a quick chat with Jordan about this on chat: we can turn off contextual logging unconditionally in `InitLogs` and override that based on a feature gate check in `ValidateAndApply`. That means the feature can only be used in some commands, but is sufficient because it includes all of the core commands.

Once the feature graduates to GA, we can enable it for all commands.

Reply:
I've pushed an update.